src/pwm_input.rs

```diff
@@ -0,0 +1,373 @@
+
+
+            // Disable capture on both channels during setting
+            // (for Channel X bit is CCXE)
+            tim.ccer.modify(|_,w| w/*.cc1e().clear_bit().cc2e().clear_bit()*/
+                            .cc1p().clear_bit().cc2p().set_bit());
+
+
+            // Define the direction of the channel (input/output)
+            // and the used input
+            // 01: CC1 channel is configured as input, IC1 is mapped on TI1.
+            // 10: CC1 channel is configured as input, IC1 is mapped on TI2.
+            tim.ccmr1_output.modify( |_,w| unsafe {w.cc1s().bits(0b01)});
+
+            // 01: CC2 channel is configured as input, IC2 is mapped on TI2
+            // 10: CC2 channel is configured as input, IC2 is mapped on TI1
+            tim.ccmr1_output.modify( |_,w| unsafe {w.cc2s().bits(0b10)});
+
+            tim.dier.write(|w| w.cc1ie().set_bit());
+
+            // Configure slave mode control register
+            // Selects the trigger input to be used to synchronize the counter
+            // 101: Filtered Timer Input 1 (TI1FP1)
+            // ---------------------------------------
+            // Slave Mode Selection :
+            //  100: Reset Mode - Rising edge of the selected trigger input (TRGI)
+            //  reinitializes the counter and generates an update of the registers.
+            tim.smcr.modify( |_,w| unsafe {w.ts().bits(0b101).sms().bits(0b100)});
+
+            match mode {
+               Frequency(f)  => {
+                  let freq = f.into().0;
+                  let max_freq = if freq > 5 {freq/5} else {1};
+                  let (arr,presc) = compute_arr_presc(max_freq, clocks.$pclk().0);
+                  tim.arr.write(|w| w.arr().bits(arr));
+                  tim.psc.write(|w| unsafe {w.psc().bits(presc)});
+
+               },
+               DutyCycle(f) => {
+                  let freq = f.into().0;
+                  let max_freq = if freq > 2 {freq/2 + freq/4 + freq/8} else {1};
+                  let (arr,presc) = compute_arr_presc(max_freq, clocks.$pclk().0);
+                  tim.arr.write(|w| w.arr().bits(arr));
+                  tim.psc.write(|w| unsafe {w.psc().bits(presc)});
+               },
+               RawFrequency(f) => {
+                  let freq = f.into().0;
+                  let (arr,presc) = compute_arr_presc(freq, clocks.$pclk().0);
+                  tim.arr.write(|w| w.arr().bits(arr));
+                  tim.psc.write(|w| unsafe {w.psc().bits(presc)});
+               }
+               RawValues{arr, presc} => {
+                  tim.arr.write(|w| w.arr().bits(arr));
+                  tim.psc.write(|w| unsafe {w.psc().bits(presc)});
+
+               }
+            }
+
+            // Enable Capture on both channels
+            tim.ccer.modify(|_,w| w.cc1e().set_bit().cc2e().set_bit());
+
+
+            tim.cr1.modify(|_,w| w.cen().set_bit());
+
+
+            unsafe { mem::uninitialized() }
```

Comment:
This is the same behavior as for the [pwm](https://github.com/stm32-rs/stm32f1xx-hal/blob/master/src/pwm.rs) module.

As far as I understand, the idea is that the struct holds no data, and since everything is implemented through macros we can use `$TIMX` to have access to the pointer block.
I might be wrong because I am not an expert with `PhantomData` !